sdks/java/extensions/sql/jdbc/src/test/java/org/apache/beam/sdk/extensions/sql/jdbc/BeamSqlLineIT.java

```diff
@@ -0,0 +1,328 @@
+                  return toLines(outputStream);
+                });
+
+    // LIMIT 2 is supposed to be parsed into 6 lines.
+    assertTrue(expectedResult.get().size() == 6);
+  }
+
+  @Test
+  public void testSlidingWindow() throws Exception {
+    ExecutorService pool = Executors.newFixedThreadPool(1);
+
+    Future<List<List<String>>> expectedResult =
+        pool.submit(
+            (Callable)
+                () -> {
+                  String[] args =
+                      buildArgs(
+                          String.format(createPubsubTableStatement, publicTopic),
+                          setProject,
+                          "SELECT COUNT(*) AS num_events,\n"
+                              + "              SUM(taxi_rides.payload.meter_increment) as revenue,\n"
+                              + "              HOP_END(\n"
+                              + "                 taxi_rides.event_timestamp, \n"
+                              + "                 INTERVAL '1' SECOND, \n"
+                              + "                 INTERVAL '2' SECOND) as minute_end\n"
+                              + "              FROM taxi_rides\n"
+                              + "              GROUP BY HOP(\n"
+                              + "                 taxi_rides.event_timestamp,\n"
+                              + "                 INTERVAL '1' SECOND, \n"
+                              + "                 INTERVAL '2' SECOND) LIMIT 2");
+                  ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+                  BeamSqlLine.runSqlLine(args, null, outputStream, null);
+                  return toLines(outputStream);
+                });
+
+    // LIMIT 2 is supposed to be parsed into 6 lines.
+    assertTrue(expectedResult.get().size() == 6);
+  }
+
+  /**
+   * Check if topic exists {@param attempts} times. There is 1 sec wait time between two attempts.
+   *
+   * @param attempts number of attempts. Has to be a value > 0
+   */
+  private void checkIfTopicExists(TestPubsub eventsTopic, int attempts)
+      throws InterruptedException, IllegalArgumentException, IOException {
+    if (attempts <= 0) {
+      throw new IllegalArgumentException(
+          String.format("Set %d attempts, which should be > 0", attempts));
+    }
+
+    while (attempts > 0) {
+      List<SubscriptionPath> topics =
+          eventsTopic.listSubscriptions(
+              projectPathFromPath(String.format("projects/%s", project)), eventsTopic.topicPath());
+      if (topics.size() > 0) {
+        return;
+      }
+
+      attempts--;
+      // Sleep 1 sec
+      Thread.sleep(1000);
+    }
+
+    throw new RuntimeException(
+        String.format("Failed to create subscription within %d attempts", attempts));
+  }
+
+  private long convertTimestampToMillis(String timestamp) throws ParseException {
+    SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```

Comment:
nit: make this a static field? E.g. `private static final DateFormat DATE_FORMAT = dateFormatInUTC("yyyy...");`

Reply:
You are right. Avoiding creating many instances of DataFormat is a big optimization.